design.md

```diff
@@ -1,16 +1,79 @@
 # AP4K Design
 
-## Terminology
+## Overview
+This section provides a high level overview on the design of AP4K. The core building blocks are back-quoted and a definition
+for them is provided in the next section (vocabulary).
 
-The section below descibes the core parts of AP4K and tries to describe how they work together.
+AP4K provides multiple annotation processors all targeting at the generation of Kubernetes resources.
+Each annotation processor is "supporting" one or more annotations and will be called by the compiler at least twice when
+the annotations are spotted. 
+
+On each annotation processor invocation, information about the annotated classes are passed.
+The last invocation signals the end of the processing.
+
+When the processing is over, each processor will assemble a `config` object and pass it to the responsible `handler`.
+The handlers role is to create or modify the `model` based on the `config`. 
+
+Finally, the `model` is serialized to disk as json/yml.
+
+[sequence diagram][https://raw.githubusercontent.com/ap4k/ap4k/master/doc/src/main/resources/sequence.png]
+
+## Vocabulary
+
+The section below describes the core parts of AP4K and tries to describe how they work together.
+
+### Visitor
+Refers to the the `Gang of Four` visitor pattern. As Kubernetes resources are deeply nested we are extensively using the 
+`visitor` pattern  to perform modifications to those resources without having to programmatically traverse this complex
+structure.
+
+### Model
+With kubernetes/openshift `model` or just `model` we refer to the java object representation of the Kubernetes/Openshift
```

Comment:
I propose to rephrase this sentence as `A `model` refers to java objects which describes the Kubernetes, Custom Resource Definition or OpenShift resources/objects.`